### PR TITLE
hann_window_torch

### LINKED ISSUE #19487: hann_window

### PR DESCRIPTION
**Pull request by ManoBharathi93**
Close #19487

### CODE REVIEW

File: ivy/functional/frontends/torch/spectral_ops.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+    window_length,
+    periodic=True,
+    *,
+    dtype=None,
+    device,
+    layout=ivy.strides,
+    required_grad=False
+) -> ivy.Array:
+    return ivy.hann_window(
+        window_length, periodic, dtype, layout, device, required_grad
```
Comment (KevinUli) [inline ivy/functional/frontends/torch/spectral_ops.py:18]: I don't know if I missed something but `ivy.hann_window()` does not seem to support `device` and `required_grad`

File: ivy_tests/test_ivy/test_frontends/test_torch/test_spectral_ops.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    ),
+    periodic=st.booleans(),
+    dtype=helpers.get_dtypes("numeric", full=False, none=True),
+    requires_grad=_requires_grad(),
+)
+def test_torch_hann_window(
+    periodic,
+    dtype_and_x,
+    dtype,
+    num_positional_args: 2,
```
Comment (KevinUli) [inline ivy_tests/test_ivy/test_frontends/test_torch/test_spectral_ops.py:20]: `num_positional_args` does not seem to be used

File: ivy_tests/test_ivy/test_frontends/test_torch/test_spectral_ops.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    on_device,
+    frontend,
+    requires_grad,
+):
+    input_dtype, x = dtype_and_x
+    helpers.test_frontend_function(
```
Comment (KevinUli) [inline ivy_tests/test_ivy/test_frontends/test_torch/test_spectral_ops.py:26]: There seems to be missing kwargs for `test_flags` and `fn_tree`

File: ivy_tests/test_ivy/test_frontends/test_torch/test_spectral_ops.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+        input_dtypes=input_dtype,
+        frontend=frontend,
+        frontend=frontend,
```
Comment (KevinUli) [inline ivy_tests/test_ivy/test_frontends/test_torch/test_spectral_ops.py:29]: There should not be multiple `frontend` kwargs

File: ivy/functional/frontends/torch/spectral_ops.py [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+
+@to_ivy_arrays_and_back
+def hann_window(
+    window_length,
+    periodic=True,
+    *,
+    dtype=None,
+    device,
+    layout=ivy.strides,
+    required_grad=False
```
Comment (KevinUli) [inline ivy/functional/frontends/torch/spectral_ops.py:15]: The kwarg `required_grad` should be `requires_grad`

**Comment by KevinUli** [comment]
Hi @ManoBharathi93,

Thanks for the PR. There seems to be a few small issues that I pointed out below. Please request a re-review after the fix.

File: ivy/functional/frontends/torch/spectral_ops.py (hunk 5 above)
Comment (KevinUli) [inline ivy/functional/frontends/torch/spectral_ops.py:15]: The kwarg required_grad should be requires_grad

File: ivy/functional/frontends/torch/spectral_ops.py (hunk 1 above)
Comment (KevinUli) [inline ivy/functional/frontends/torch/spectral_ops.py:18]: I don't know if I missed something but ivy.hann_window() does not seem to support device and required_grad

File: ivy_tests/test_ivy/test_frontends/test_torch/test_spectral_ops.py (hunk 2 above)
Comment (KevinUli) [inline ivy_tests/test_ivy/test_frontends/test_torch/test_spectral_ops.py:20]: `num_positional_args` does not seem to be used

File: ivy_tests/test_ivy/test_frontends/test_torch/test_spectral_ops.py (hunk 3 above)
Comment (KevinUli) [inline ivy_tests/test_ivy/test_frontends/test_torch/test_spectral_ops.py:26]: There seems to be missing kwargs for `test_flags` and `fn_tree`

File: ivy_tests/test_ivy/test_frontends/test_torch/test_spectral_ops.py (hunk 4 above)
Comment (KevinUli) [inline ivy_tests/test_ivy/test_frontends/test_torch/test_spectral_ops.py:29]: There should not be multiple `frontend` kwargs

File: ivy/functional/frontends/torch/spectral_ops.py (hunk 1 above)
Comment (KevinUli) [inline ivy/functional/frontends/torch/spectral_ops.py:18]: I don't know if I missed something but `ivy.hann_window()` does not seem to support `device` and `required_grad`

File: ivy/functional/frontends/torch/spectral_ops.py (hunk 5 above)
Comment (KevinUli) [inline ivy/functional/frontends/torch/spectral_ops.py:15]: The kwarg `required_grad` should be `requires_grad`

File: ivy_tests/test_ivy/test_frontends/test_torch/test_spectral_ops.py (hunk 2 above)
Comment (KevinUli) [inline ivy_tests/test_ivy/test_frontends/test_torch/test_spectral_ops.py:20]: `num_positional_args` does not seem to be used

File: ivy_tests/test_ivy/test_frontends/test_torch/test_spectral_ops.py (hunk 3 above)
Comment (KevinUli) [inline ivy_tests/test_ivy/test_frontends/test_torch/test_spectral_ops.py:26]: There seems to be missing kwargs for `test_flags` and `fn_tree`

File: ivy_tests/test_ivy/test_frontends/test_torch/test_spectral_ops.py [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    dtype,
+    num_positional_args: 2,
+    on_device,
+    frontend,
+    requires_grad,
+):
+    input_dtype, x = dtype_and_x
+    helpers.test_frontend_function(
+        input_dtypes=input_dtype,
+        frontend=frontend,
```
Comment (KevinUli) [inline ivy_tests/test_ivy/test_frontends/test_torch/test_spectral_ops.py:28]: There should not be multiple `frontend` kwargs

File: ivy/functional/frontends/torch/spectral_ops.py (hunk 1 above)
Comment (ManoBharathi93) [reply ivy/functional/frontends/torch/spectral_ops.py:18]: I think you have missed device and required_grad... In backend of torch it's a kwargs..

https://pytorch.org/docs/stable/generated/torch.hann_window.html

File: ivy_tests/test_ivy/test_frontends/test_torch/test_spectral_ops.py (hunk 6 above)
Comment (ManoBharathi93) [reply ivy_tests/test_ivy/test_frontends/test_torch/test_spectral_ops.py:16]: Yes, I will remove it....Thanks for pointing out it.

File: ivy_tests/test_ivy/test_frontends/test_torch/test_spectral_ops.py (hunk 3 above)
Comment (ManoBharathi93) [reply ivy_tests/test_ivy/test_frontends/test_torch/test_spectral_ops.py:14]: yes, I missed it. i.e: `fn_tree` & `test_flags` need to  pass the generated values to `helpers.test_frontend_function` 

sorry to bother you so much....

File: ivy_tests/test_ivy/test_frontends/test_torch/test_spectral_ops.py (hunk 2 above)
Comment (ManoBharathi93) [reply ivy_tests/test_ivy/test_frontends/test_torch/test_spectral_ops.py]: removed it😁

File: ivy/functional/frontends/torch/spectral_ops.py (hunk 5 above)
Comment (ManoBharathi93) [reply ivy/functional/frontends/torch/spectral_ops.py]: I think we don't need it.. As the backend of ivy doesn't  require `requires_grad`

**Comment by KevinUli** [comment]
Sorry for the confusion. The torch frontend function implementation should have the parameters of the torch framework [here](https://pytorch.org/docs/stable/generated/torch.hann_window.html). As for the parameters not supported by the Ivy API, you should try to implement it using a combination of Ivy APIs

File: ivy/functional/frontends/torch/spectral_ops.py [hunk 7]
```diff
@@ -0,0 +1,11 @@
+import ivy
+from ivy.functional.frontends.torch.func_wrapper import (
+    to_ivy_arrays_and_back,
+)
+from ivy.func_wrapper import with_supported_dtypes
+
+
+@with_supported_dtypes({"2.0.1 and below": "int"}, "torch")
+@to_ivy_arrays_and_back
+def hann_window(size, *, periodic=True, dtype=None, out=None) -> ivy.Array:
```
Comment (KevinUli) [inline ivy/functional/frontends/torch/spectral_ops.py:10]: This should stay the same as your previous implementation
```suggestion
def hann_window(window_length, periodic=True, *, dtype=None, layout=ivy.strided, device=None, requires_grad=False):
```

**Comment by ManoBharathi93** [comment]
> Sorry for the confusion. The torch frontend function implementation should have the parameters of the torch framework [here](https://pytorch.org/docs/stable/generated/torch.hann_window.html). As for the parameters not supported by the Ivy API, you should try to implement it using a combination of Ivy APIs

it's okay, I take your suggestions and committed. Kindly review it once again if you find any other issue pls mention it..

**Comment by KevinUli** [comment]
Please run the pre-commit shown [here](https://unify.ai/docs/ivy/overview/deep_dive/formatting.html) to make sure the formatting is correct. Please also run the test and get it to run successfully. You can find some instructions on how the frontend implementation works [here](https://unify.ai/docs/ivy/overview/deep_dive/ivy_frontends.html), and instructions on how the testing of the frontend works [here](https://unify.ai/docs/ivy/overview/deep_dive/ivy_frontends_tests.html)

File: ivy/functional/frontends/torch/spectral_ops.py
Comment (KevinUli) [inline ivy/functional/frontends/torch/spectral_ops.py:1]: One problem why test is not running is due to `ivy.strided` not being implemented as an Ivy API. You might want to consider implementing the `layout` kwarg in the frontend using `ivy.strides`.

Comment (ManoBharathi93) [reply ivy/functional/frontends/torch/spectral_ops.py:1]: https://github.com/unifyai/ivy/blob/a5c873b2b826c855918dfb1a513d00966999f947/ivy/functional/ivy/general.py#L4159

`ivy.strided` is implemented in Ivy API...you can refer the above to the above link..

**Comment by ManoBharathi93** [comment]
> 

sorry I just committed you suggested commit directly on Github that's the problem...

**Comment by ManoBharathi93** [comment]
This test cases are passing in my local run.. but failing here!

I think problem is that device kwarg is not supporting in `ivy api`

**Comment by ManoBharathi93** [comment]
@KevinUli  Unfortunately, I stuck with development environment issue... I am trying to fix and asked for help in our discord community bro.. (I couldn't run any test in my vscode ) 
image
I tried edit configuration and manually added the target but it's not working for me...

File: ivy/functional/frontends/torch/spectral_ops.py [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+from ivy.functional.frontends.torch.func_wrapper import (
+    to_ivy_arrays_and_back,
+)
+from ivy.func_wrapper import with_supported_dtypes
+
+
+@with_supported_dtypes({"2.0.1 and below": "int"}, "torch")
+@to_ivy_arrays_and_back
+def hann_window(
+    window_length, periodic=True, *, dtype=None, requires_grad=False, out=None
```
Comment (KevinUli) [inline ivy/functional/frontends/torch/spectral_ops.py:11]: The parameters for `hann_window` torch frontend should be the same as torch's as shown [here](https://pytorch.org/docs/stable/generated/torch.hann_window.html). It seems to be missing `layout` and `device` kwarg

Comment (ManoBharathi93) [reply ivy/functional/frontends/torch/spectral_ops.py:11]: Yes, that what I have implemented initially but it fails test case..

File: ivy/functional/frontends/torch/spectral_ops.py [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+):
+    return ivy.hann_window(
+        window_length, periodic, dtype, requires_grad=requires_grad, out=out
```
Comment (KevinUli) [inline ivy/functional/frontends/torch/spectral_ops.py:14]: `ivy.hann_window` does not seem to support `requires_grad` kwarg. `periodic` and `dtype` are keyword-only arguments, therefore it should be kwargs

File: ivy/functional/frontends/torch/spectral_ops.py [hunk 10]
```diff
@@ -0,0 +1,15 @@
+import ivy
+from ivy.functional.frontends.torch.func_wrapper import (
+    to_ivy_arrays_and_back,
+)
+from ivy.func_wrapper import with_supported_dtypes
+
+
+@with_supported_dtypes({"2.0.1 and below": "int"}, "torch")
```
Comment (KevinUli) [inline ivy/functional/frontends/torch/spectral_ops.py:8]: the value for the dict should be a tuple of supported dtypes. Ex: `("int32", "int64")` (You should check to see exactly which int dtypes are supported)

File: ivy/functional/frontends/torch/spectral_ops.py (hunk 9 above)
Comment (ManoBharathi93) [reply ivy/functional/frontends/torch/spectral_ops.py:14]: yes, I tried yesterday but it fails

File: ivy_tests/test_ivy/test_frontends/test_torch/test_spectral_ops.py [hunk 11]
```diff
@@ -0,0 +1,33 @@
+import ivy_tests.test_ivy.helpers as helpers
+from hypothesis import strategies as st
+from ivy_tests.test_ivy.helpers.testing_helpers import handle_frontend_test
+from ivy_tests.test_ivy.test_frontends.test_torch.test_tensor import _requires_grad
+
+
+@handle_frontend_test(
+    fn_tree="torch.hann_window",
+    dtype_and_x=helpers.dtype_and_values(available_dtypes=helpers.get_dtypes("int")),
```
Comment (KevinUli) [inline ivy_tests/test_ivy/test_frontends/test_torch/test_spectral_ops.py:9]: the `kind` arg for `helpers.get_dtypes` does not take `int`. It should be `integer`. (You can check the docstring to see all the values it takes)

File: ivy/functional/frontends/torch/spectral_ops.py (hunk 10 above)
Comment (ManoBharathi93) [reply ivy/functional/frontends/torch/spectral_ops.py:8]: That's right, I missed this...let me try it..

File: ivy_tests/test_ivy/test_frontends/test_torch/test_spectral_ops.py (hunk 11 above)
Comment (ManoBharathi93) [reply ivy_tests/test_ivy/test_frontends/test_torch/test_spectral_ops.py:9]: got it.. thanks

File: ivy/functional/frontends/torch/spectral_ops.py (hunk 8 above)
Comment (ManoBharathi93) [reply ivy/functional/frontends/torch/spectral_ops.py:11]: > The parameters for `hann_window` torch frontend should be the same as torch's as shown [here](https://pytorch.org/docs/stable/generated/torch.hann_window.html). It seems to be missing `layout` and `device` kwarg. It also should not have `out` parameter

image

Comment (KevinUli) [reply ivy/functional/frontends/torch/spectral_ops.py:11]: The `device` and `layout` kwarg should be in the frontend parameter but not in the `ivy.hann_window` kwargs as `ivy.hann_window` does not support them.

Comment (ManoBharathi93) [reply ivy/functional/frontends/torch/spectral_ops.py:11]: you mean, Implement `device` and `layout` kwarg  in function paramter but dont want  to use it return statement ivy.hann_window() am i right ?

Comment (KevinUli) [reply ivy/functional/frontends/torch/spectral_ops.py:11]: Right. For `layout` you might want to try `ivy.strides()` and `ivy.as_strided()`. For `device` you might want to try `ivy.to_device()`.

Comment (ManoBharathi93) [reply ivy/functional/frontends/torch/spectral_ops.py:11]: ivy.hann_windows expects floating dtype...!

**Comment by ManoBharathi93** [comment]
https://github.com/unifyai/ivy/blob/49206e13e2fd30c164bcc3f7cc7794774de0eb9c/ivy/functional/ivy/experimental/creation.py#L68
image
backend doesn't have device kwargs....

**Comment by KevinUli** [comment]
> https://github.com/unifyai/ivy/blob/49206e13e2fd30c164bcc3f7cc7794774de0eb9c/ivy/functional/ivy/experimental/creation.py#L68
> 
> 
> image
> backend doesn't have device kwargs....

I was hinting towards something like 
`def hann_window(...):
     ret = ivy.hann_window(...)
     return ivy.to_device(ret, device)`

**Comment by ManoBharathi93** [comment]
image

I am closing this pull request because the backend code implemented actually doesn't support torch frontend from my perspective.. 

Thank you

**Pull request closed without merging**